File: scripts/texture-pipeline/debug_pole.py

import math

def s2_st_to_uv(s):
    if s >= 0.5:
        return (1.0 / 3.0) * (4.0 * s * s - 1.0)
    else:
        return (1.0 / 3.0) * (1.0 - 4.0 * (1.0 - s) * (1.0 - s))

def face_uv_to_xyz(face, u, v):
    su = s2_st_to_uv(u)
    sv = s2_st_to_uv(v)
    
    if face == 2:
        x = -su
        y = -sv
        z = 1.0
    elif face == 5:
        x = sv
        y = su
        z = -1.0
    else:
        return 0,0,0

    r = math.sqrt(x*x + y*y + z*z)
    return (x/r, y/r, z/r)

def xyz_to_latlon(x, y, z):
    # z is sin(lat) if spherical
    # clamp z
    z = max(-1.0, min(1.0, z))
    lat = math.asin(z) * 180.0 / math.pi
    lon = math.atan2(y, x) * 180.0 / math.pi
    return lat, lon

def main():
    W = 1024
    offset = 0.5
    center = W // 2
    
    print(f"Debugging Face 2 Center (W={W})...")
    print(f"{'row':<5} | {'col':<5} | {'u':<10} | {'v':<10} | {'x':<10} | {'y':<10} | {'z':<10} | {'lat':<10} | {'lon':<10}")
    
    for r in range(center - 2, center + 3):
        for c in range(center - 2, center + 3):
            u = (c + offset) / W
            v = 1.0 - (r + offset) / W
            
            x, y, z = face_uv_to_xyz(2, u, v)
            lat, lon = xyz_to_latlon(x, y, z)
            
            print(f"{r:<5} | {c:<5} | {u:<10.6f} | {v:<10.6f} | {x:<10.6f} | {y:<10.6f} | {z:<10.6f} | {lat:<10.4f} | {lon:<10.4f}")

if __name__ == "__main__":
    main()

File: scripts/debug/debug_pole.cpp

#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>
#include <iomanip>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

// Copy of S2 Logic
inline double s2_st_to_uv(double s) {
    if (s >= 0.5) return (1.0 / 3.0) * (4.0 * s * s - 1.0);
    return (1.0 / 3.0) * (1.0 - 4.0 * (1.0 - s) * (1.0 - s));
}

struct Point3D { double x, y, z; };

Point3D face_uv_to_xyz(int face, double u, double v) {
    double su = s2_st_to_uv(u);
    double sv = s2_st_to_uv(v);
    double x, y, z;
    switch (face) {
        case 0: x =  1.0; y =   su; z =   sv; break; 
        case 1: x = -su;  y =  1.0; z =   sv; break; 
        case 2: x = -su;  y = -sv;  z =  1.0; break; 
        case 3: x = -1.0; y = -sv;  z = -su;  break; 
        case 4: x =  sv;  y = -1.0; z = -su;  break; 
        case 5: x =  sv;  y =   su; z = -1.0; break; 
        default: x = y = z = 0; break;
    }
    double r = std::sqrt(x * x + y * y + z * z);
    return { x / r, y / r, z / r };
}

void xyz_to_latlon(const Point3D& p, double& lat, double& lon) {
    lon = std::atan2(p.y, p.x) * 180.0 / M_PI;
    lat = std::asin(std::max(-1.0, std::min(1.0, p.z))) * 180.0 / M_PI;
}

int main() {
    int W = 1024;
    double offset = 0.5; // Pixel centered
    int center = W / 2;
    
    std::cout << "Debugging Face 2 Center (W=" << W << ")..." << std::endl;
    std::cout << "row | col | u | v | x | y | z | lat | lon" << std::endl;
    
    // Check 3x3 around center
    for (int r = center - 1; r <= center + 1; ++r) {
        for (int c = center - 1; c <= center + 1; ++c) {
            double u = (double)(c + offset) / (double)W;
            double v = 1.0 - (double)(r + offset) / (double)W;
            
            Point3D p = face_uv_to_xyz(2, u, v);
            double lat, lon;
            xyz_to_latlon(p, lat, lon);
            
            std::cout << r << " | " << c << " | " 
                      << std::fixed << std::setprecision(6) << u << " | " << v << " | "
                      << p.x << " | " << p.y << " | " << p.z << " | "
                      << lat << " | " << lon << std::endl;
        }
    }
    
    return 0;
}
